hydroflow/src/builder/build/pull_fold_epoch.rs

```diff
@@ -0,0 +1,53 @@
+use super::{PullBuild, PullBuildBase};
+
+use crate::scheduled::{context::Context, handoff::handoff_list::PortList, port::RECV};
+
+pub struct FoldEpochPullBuild<Prev, Init, Func>
+where
+    Prev: PullBuild,
+{
+    prev: Prev,
+    init: Init,
+    func: Func,
+}
+impl<Prev, Init, Func, Out> FoldEpochPullBuild<Prev, Init, Func>
+where
+    Prev: PullBuild,
+    Init: FnMut(&Context<'_>) -> Out,
+    Func: FnMut(&Context<'_>, Out, Prev::ItemOut) -> Out,
+{
+    pub fn new(prev: Prev, init: Init, func: Func) -> Self {
+        Self { prev, init, func }
+    }
+}
+
+impl<Prev, Init, Func, Out> PullBuildBase for FoldEpochPullBuild<Prev, Init, Func>
+where
+    Prev: PullBuild,
+    Init: FnMut(&Context<'_>) -> Out,
+    Func: FnMut(&Context<'_>, Out, Prev::ItemOut) -> Out,
+{
+    type ItemOut = Out;
+    type Build<'slf, 'ctx> = std::iter::Once<Out>;
+}
+
+impl<Prev, Init, Func, Out> PullBuild for FoldEpochPullBuild<Prev, Init, Func>
+where
+    Prev: PullBuild,
+    Init: FnMut(&Context<'_>) -> Out,
+    Func: FnMut(&Context<'_>, Out, Prev::ItemOut) -> Out,
+{
+    type InputHandoffs = Prev::InputHandoffs;
+
+    fn build<'slf, 'ctx>(
+        &'slf mut self,
+        context: &'ctx Context<'ctx>,
+        handoffs: <Self::InputHandoffs as PortList<RECV>>::Ctx<'ctx>,
+    ) -> Self::Build<'slf, 'ctx> {
+        std::iter::once(
+            self.prev
```

Comment:
It feels slightly weird to me that this does its work in the build phase rather than the actual execution phase, probably doesn't semantically matter? But for metrics reasons it might be nice to keep those two phases separate? Not really sure. Don't have a strong opinion.

Reply:
Good catch